File: app/components/folio/console/files/index_modal_component.rb
# frozen_string_literal: true

class Folio::Console::Files::IndexModalComponent < Folio::Console::ApplicationComponent
  CLASS_NAME = "f-c-files-index-modal"

  def initialize
  end

  def data
    stimulus_controller(CLASS_NAME,
                        values: { turbo_frame_config: },
                        action: {
                          "f-c-files-index-modal:openWithType" => "openWithType",
                          "f-file-list-file:select" => "onFileSelect",
                          "f-modal:closed" => "onModalClosed",
                        })
  end

  def turbo_frame_config
    h = {}

    Rails.application.config.folio_file_types_for_routes.each do |type|
      klass = type.constantize
      h[type] = {
        src: url_for([:console, klass]),
        id: klass.console_turbo_frame_id(modal: true),
      }
    end

    h
  end

  def file_klasses
    Rails.application.config.folio_file_types_for_routes.map(&:constantize)
  end
end

File: app/components/folio/console/files/index_modal_component.js
window.Folio.Stimulus.register('f-c-files-index-modal', class extends window.Stimulus.Controller {
  static values = {
    turboFrameConfig: Object
  }

  static targets = ['type']

  disconnect () {
    delete this.triggerElement
  }

  openWithType (e) {
    if (!e.detail.fileType) {
      window.alert('[Folio::Console::Files::IndexModalComponent] Missing fileType!')
      return
    }

    if (!e.detail.trigger) {
      window.alert('[Folio::Console::Files::IndexModalComponent] Missing trigger element!')
      return
    }

    this.triggerElement = e.detail.trigger

    let typeElement

    this.typeTargets.forEach((typeTarget) => {
      const turboFrame = typeTarget.querySelector('turbo-frame')

      if (typeTarget.dataset.type === e.detail.fileType) {
        if (turboFrame.disabled) {
          turboFrame.disabled = false
        }

        typeTarget.hidden = false

        typeElement = typeTarget
      } else {
        if (!turboFrame.disabled) {
          turboFrame.disabled = true
        }

        typeTarget.hidden = true
      }
    })

    if (!typeElement) {
      window.alert(`[Folio::Console::Files::IndexModalComponent] Invalid fileType: ${e.detail.fileType}!`)
      return
    }

    window.Folio.Modal.open(this.element)
  }

  onFileSelect (e) {
    if (!this.triggerElement) {
      window.alert('[Folio::Console::Files::IndexModalComponent] Missing trigger element!')
      return
    }

    if (!e.detail.fileId) {
      window.alert('[Folio::Console::Files::IndexModalComponent] Missing fileId in event detail!')
      return
    }

    this.triggerElement.dispatchEvent(new CustomEvent('f-c-files-index-modal:selectedFile', {
      detail: {
        fileId: e.detail.fileId
      }
    }))

    window.Folio.Modal.close(this.element)
  }

  onModalClosed () {
    delete this.triggerElement
  }
})
